types.c

```diff
@@ -1623,8 +1623,11 @@ ucv_to_stringbuf_formatted(uc_vm_t *vm, uc_stringbuf_t *pb, uc_value_t *uv, size
 			ucv_stringbuf_append(pb, "Infinity");
 		else if (d == -INFINITY)
 			ucv_stringbuf_append(pb, "-Infinity");
-		else
-			ucv_stringbuf_printf(pb, "%.14g", d);
+		else {
+			size_t len = ucv_stringbuf_printf(pb, "%.14g", d);
+			if (json && !strchr(pb->buf + pb->bpos - len, '.'))
```

Comment:
Hmm, this feels slightly unelegant. What about this:

```c
...
else if (json && trunc(d) == d)
    ucv_stringbuf_printf(pb, "%.1f", d);
else 
    ucv_stringbuf_printf(pb, "%.14g", d);
```

Reply:
I thought about doing it this way too, but I read in a few places that this can be prone to corner cases caused by weird floating point rounding issue. I think checking the output of sprintf is more reliable.

Reply:
> can be prone to corner cases caused by weird floating point rounding issue

Hm, can you provide some links for this? My research so far seems to indicate that `trunc(x) == x` is indeed the way to go, but maybe I overlooked something.

The problem with the "scan backwards for a decimal point and append `.0` if none found" approach is that it will fail if the `%g` format decides to yield an exponent form, e.g. we don't want to turn `1e100` into `1e100.0`.

Reply:
Sorry, I can't find the original link anymore. I took a look at this from another angle and checked how existing JSON libraries are handling this issue. I looked at json-c, jansson and yajl, and they all use the scan-for-decimal-point approach. The code is rather different in each one, so I don't believe it was copied/cargo-culted between implementations. Would you accept my patch if I keep the approach, but fix the exponent form case?

Reply:
I pushed a fix based on your suggestion and including a testcase covering this corner case. Could you rebase this PR against master so that I can merge the remainder?

Reply:
Done